### PR TITLE
Changes to enable GPU gear execution through `gpu` tag

### PR DESCRIPTION
**Pull request by joshicola**
This PR introduces SBATCH file and fw-cast changes to enable GPU execution of gears on an appropriately configured Slurm Cluster.

### CODE REVIEW

File: doc/faq_use_gpu_on_slurm_cluster.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  - [Problems](#problems)
+
+## Configuration
+
+Both Slurm and fw-cast must be configured appropriately to enable GPU execution of gears
+on a Slurm Cluster.
+
+### Slurm Configuration
+
+To execute gears on GPU nodes of a Slurm cluster, the slurm must be configured
```
Comment (jiavila) [inline doc/faq_use_gpu_on_slurm_cluster.md]:
```suggestion
To execute gears on GPU nodes of a Slurm cluster, Slurm must be configured
```

File: doc/faq_use_gpu_on_slurm_cluster.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+The [`slurm.conf`](https://slurm.schedmd.com/slurm.conf.html) file is typically found in
+`/etc/slurm/`. Below is an example for a node definition that enables GPUs to be
+scheduled.
+
+```conf
+NodeName=scien-hpc-gpu Gres=gpu:tesla:1 CPUs=4 Boards=1 SocketsPerBoard=1 CoresPerSocket=2 ThreadsPerCore=2 RealMemory=14978
+```
+
+The Generic RESource (GRES) flag (`Gres=gpu:tesla:1`) must be present to indicate the
+resource type ("gpu"), the resource class ("tesla"), and the number of resources present
```
Comment (jiavila) [inline doc/faq_use_gpu_on_slurm_cluster.md]:
```suggestion
resource type (e.g., "gpu"), the resource class (e.g., "tesla"), and the number of resources present
```
does this have to be "gpu" and "tesla" If not, then suggest adding e.g. to indicate it's an example

File: doc/faq_use_gpu_on_slurm_cluster.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+("1"). Execution on more than one GPU per node has not yet been explored.
+
+The remainder of the node configuration (e.g. CPUs, RealMemory) can be interrogated by
```
Comment (jiavila) [inline doc/faq_use_gpu_on_slurm_cluster.md]:
```suggestion
FYI: the remainder of the node configuration (e.g. CPUs, RealMemory) can be interrogated by
```
to make it clear that they don't have to do anything with this

File: doc/faq_use_gpu_on_slurm_cluster.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+the following command:
+
+```bash
+slurmd -C
+```
+
+#### gres.conf
+
+The [Generic RESource (GRES)](https://slurm.schedmd.com/gres.conf.html) configuration,
+`gres.conf` needs to have an entry for each resource named in `slurm.conf`.
```
Comment (jiavila) [inline doc/faq_use_gpu_on_slurm_cluster.md]:
```suggestion
`gres.conf`, needs to have an entry for each resource named in `slurm.conf`.
```

File: doc/faq_use_gpu_on_slurm_cluster.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+   srun ./engine run --single-job {{job.fw_id}}
+```
+
+## Gear Execution
+
+With the rest of the workflow configured, adding a `gpu` tag (in addition to the
+`hpc` tag) to the launch of the gear will schedule a GPU to execute the gear on the
+Slurm cluster.
+
+### Problems
```
Comment (jiavila) [inline doc/faq_use_gpu_on_slurm_cluster.md]:
```suggestion
### Potential Problems
```

File: pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,27 @@
+[tool.poetry]
```
Comment (jiavila) [inline pyproject.toml]: I know that currently one has to cd in to the src code and then use pipenv to create the environment. I'm assuming that you're going to remove all the associated pipfiles, then change [these instructions](https://github.com/joshicola/hpc-client-josh/blob/gpu_changes/contributing.md) and [these instructions](https://github.com/joshicola/hpc-client-josh/blob/gpu_changes/doc/3-cluster-install.md#clone-the-cast-repo-and-set-up-a-virtual-environment) once we convert the hpc-client into a standalone package. Will users still have to install the entire repo, since they still need to perform the setup with the examples from the package?

Comment (jiavila) [reply pyproject.toml]: FWIW, kofalt was kinda indifferent between poetry and pipenv ("what more does poetry give you? seriously?" :D)

File: src/cluster/__init__.py [hunk 7]
```diff
@@ -1,4 +1,4 @@
-from util import frame
```
Comment (jiavila) [inline src/cluster/__init__.py]: the pipenv is inside the src folder, so the reference has to remain (unless we provide instructions for managing the env with poetry, like I mentioned in previous comment).

File: src/cluster/base.py [hunk 8]
```diff
@@ -7,239 +7,308 @@
 
 
 class Base(Common):
-	"""
```
Comment (jiavila) [inline src/cluster/base.py:10]: For some reason, indentations were finicky when I was developing...I left it as is to not break it. I'm assuming this fixes everything, since you've tested the entire installation from scratch?

File: src/cluster/common.py [hunk 9]
```diff
@@ -213,68 +213,3 @@ def report_results(self, start, jobs_launched, jobs_skipped, jobs_rejected):
 
         self.log.info(msg)
 
-    def handle_all(self, start):
```
Comment (jiavila) [inline src/cluster/common.py:216]: Why was this moved? I'm assuming Kofalt designed it this way for abstraction/encapsulation.

File: src/cluster/slurm.py [hunk 10]
```diff
@@ -5,50 +5,53 @@
```
Comment (jiavila) [inline src/cluster/slurm.py:5]: same thing about spacing as comment above 🤞

File: src/util/net.py [hunk 11]
```diff
@@ -55,10 +55,14 @@ def prepare_search(cast_config):
 
 
 def load_user_id_whitelist(fw):
-	"""
-	Load user IDs from a FW-group-defined whitelist.
-	"""
+	"""Load user IDs from a FW-group-defined whitelist.
+
+	Args:
+		fw (flywheel.Client): Flywheel client instance.
```
Comment (jiavila) [inline src/util/net.py:61]: nice doc, bro. I hated going through this thing the first time under pressure and trying to figure all the objects

File: tests/util/test_defn.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
+import pytest
+from src.util.defn import JobSettings
+from pydantic import ValidationError
+
+def test_job_settings_creation():
+    job_settings_data = {
```
Comment (jiavila) [inline tests/util/test_defn.py:6]: One thing that made sense from Suzanne was to add docstrings to the test methods, just to make sure it's clear for all levels of users/developers and when we revisit this test years later. I recommend adding if you have time. Thanks!

**Review by jiavila — CHANGES_REQUESTED** [review]
Please see comments about potentially breaking changes. Thanks!

File: src/cluster/base.py (hunk 8 above)
Comment (joshicola) [reply src/cluster/base.py:10]: This file was `black` formatted to address those finicky indentations in a consistent manner.

It functions as expected in an installation on `scien-hpc`.

Comment (joshicola) [reply src/cluster/base.py:10]: In a future PR, I will apply `black` and `isort` across all Python files in this repository.

File: src/cluster/common.py (hunk 9 above)
Comment (joshicola) [reply src/cluster/common.py:216]: `handle_all` was moved because the functions it references are in the descendant class (Base) and make it harder to test directly.

This is a non-breaking change as demonstrated in tests on `scien-hpc`.

File: src/cluster/slurm.py (hunk 10 above)
Comment (joshicola) [reply src/cluster/slurm.py:5]: Yes, thank you for being a stickler.

`black` was applied to this file for conformance to the SSE coding standards.

Expect a dedicated PR in the near future exclusively dedicated to formatting.

File: src/util/net.py (hunk 11 above)
Comment (joshicola) [reply src/util/net.py:61]: Thanks,

This is my Jam.

Expect a future PR dedicated to updating all of the Docstrings.

File: tests/util/test_defn.py (hunk 12 above)
Comment (joshicola) [reply tests/util/test_defn.py:6]: I am going to put that in the "Docstring" PR.

File: src/cluster/__init__.py (hunk 7 above)
Comment (joshicola) [reply src/cluster/__init__.py]: As above, I agree that this `pipenv` to `poetry` transition is premature. I am punting to a future point.

**Review by jiavila — APPROVED** [review]
lgtm
